File: LeetCode75/hashmap and sets/findTheDifferenceOfTwoArrays.js
//My naive solution:
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[][]}
 */
var findDifference = function (nums1, nums2) {
  let arr1 = [];
  let arr2 = [];
  for (const val of nums1) {
    if (!nums2.includes(val)) {
      arr1.push(val);
    }
  }
  for (const val of nums2) {
    if (!nums1.includes(val)) {
      arr2.push(val);
    }
  }
  return [_.uniq(arr1), _.uniq(arr2)];
};

//optimal solution using sets:
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[][]}
 */
function findDifference(nums1, nums2) {
  // Use sets for efficient difference operations
  const set1 = new Set(nums1);
  const set2 = new Set(nums2);

  // Elements in nums1 but not in nums2
  const difference1 = [...set1].filter((x) => !set2.has(x));

  // Elements in nums2 but not in nums1
  const difference2 = [...set2].filter((x) => !set1.has(x));

  return [difference1, difference2];
}

File: LeetCode75/hashmap and sets/findTheDifferenceOfTwoArrays.py
# Given two 0-indexed integer arrays nums1 and nums2, return a list answer of size 2 where:

# answer[0] is a list of all distinct integers in nums1 which are not present in nums2.
# answer[1] is a list of all distinct integers in nums2 which are not present in nums1.
# Note that the integers in the lists may be returned in any order.

class Solution(object):
    def findDifference(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[List[int]]
        """
        set1, set2 = set(nums1), set(nums2)
        arr1, arr2 = set(), set()

        for val in set1:
            if val not in set2:
                arr1.add(val)
        
        for val in set2:
            if val not in set1:
                arr2.add(val)
        
        return [list(arr1), list(arr2)]

        